plugins/corsair/fu-corsair-bp.c

```diff
@@ -0,0 +1,407 @@
+	if (!fu_memcpy_safe(cmd,
+			    sizeof(cmd),
+			    CORSAIR_NEXT_CHUNKS_HEADER_SIZE,
+			    fu_chunk_get_data(chunk),
+			    fu_chunk_get_data_sz(chunk),
+			    0,
+			    fu_chunk_get_data_sz(chunk),
+			    error)) {
+		g_prefix_error(error, "cannot set data: ");
+		return FALSE;
+	}
+	if (!fu_corsair_bp_command(self, cmd, CORSAIR_TRANSACTION_TIMEOUT, TRUE, error)) {
+		g_prefix_error(error, "write command fail: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static void
+fu_corsair_bp_class_init(FuCorsairBpClass *klass)
+{
+}
+
+static void
+fu_corsair_bp_init(FuCorsairBp *self)
+{
+	self->cmd_read_size = CORSAIR_DEFAULT_CMD_SIZE;
+	self->cmd_write_size = CORSAIR_DEFAULT_CMD_SIZE;
+	self->is_subdevice = FALSE;
+}
+
+gboolean
+fu_corsair_bp_get_property(FuCorsairBp *self,
+			   FuCorsairBpProperty property,
+			   guint32 *value,
+			   GError **error)
+{
+	guint8 data[FU_CORSAIR_MAX_CMD_SIZE] = {0x08, 0x02};
+
+	fu_common_write_uint16(&data[CORSAIR_OFFSET_CMD_PROPERTY_ID],
+			       (guint16)property,
+			       G_LITTLE_ENDIAN);
+
+	if (!fu_corsair_bp_command(self, data, CORSAIR_TRANSACTION_TIMEOUT, TRUE, error))
+		return FALSE;
+
+	*value = fu_common_read_uint32(&data[CORSAIR_OFFSET_CMD_PROPERTY_VALUE], G_LITTLE_ENDIAN);
+
+	return TRUE;
+}
+
+gboolean
+fu_corsair_bp_set_mode(FuCorsairBp *self, FuCorsairDeviceMode mode, GError **error)
+{
+	guint8 cmd[FU_CORSAIR_MAX_CMD_SIZE] = {0x08, 0x01, 0x03};
+
+	cmd[CORSAIR_OFFSET_CMD_SET_MODE] = mode;
+
+	if (!fu_corsair_bp_command(self, cmd, CORSAIR_TRANSACTION_TIMEOUT, TRUE, error)) {
+		g_prefix_error(error, "set mode command fail: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_corsair_bp_write_firmware_chunks(FuCorsairBp *self,
+				    FuChunk *first_chunk,
+				    GPtrArray *chunks,
+				    FuProgress *progress,
+				    guint32 firmware_size,
+				    GError **error)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len + 1);
+
+	if (!fu_corsair_bp_write_first_chunk(self, first_chunk, firmware_size, error)) {
+		g_prefix_error(error, "cannot write first chunk: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	for (guint id = 0; id < chunks->len; id++) {
+		FuChunk *chunk = g_ptr_array_index(chunks, id);
+		if (!fu_corsair_bp_write_chunk(self, chunk, error)) {
+			g_prefix_error(error, "cannot write chunk %u", id);
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_corsair_bp_commit_firmware(FuCorsairBp *self, GError **error)
+{
+	guint8 commit_cmd[FU_CORSAIR_MAX_CMD_SIZE] = {0x08, 0x05, 0x01, 0x00};
+	if (!fu_corsair_bp_command(self, commit_cmd, CORSAIR_TRANSACTION_TIMEOUT, TRUE, error)) {
+		g_prefix_error(error, "firmware commit fail: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+gboolean
+fu_corsair_bp_write_firmware(FuCorsairBp *self,
+			     FuFirmware *firmware,
+			     FuProgress *progress,
+			     GError **error)
```

Comment:
if you derive from FuUsbDevice you can use the ->write_firmware() vfunc

Reply:
I think if FuCorsairBp subclasses any of the FuDevice objects (or FuDevice itself) then you can hook this up as a vfunc.

Reply:
I still think FuCorsairBp should derive from FuUsbDevice

Reply:
derived from FuUsbDevice

Reply:
Same, can you connect this vfunc and then use `fu_device_write_firmware(bp)` from the caller? Then all the logging stuff works correctly.